### ISSUE TITLE
Implement Earth mover's distance (EMD)

### ISSUE THREAD

**Issue by MartinThoma** [scipy/scipy#7804]
This is a feature request. Please implement [Earth mover's distance (EMD)](https://en.wikipedia.org/wiki/Earth_mover%27s_distance). It can be used to compare probability distributions.

Actually, it seems not too difficult to implement it. So if it is wanted by the project owners, I could do that. I guess scipy.spatial.distance would be a good place.

See [A Metric for Distributions with Applications to Image Databases](https://users.cs.duke.edu/~tomasi/papers/rubner/rubnerIccv98.pdf)

**Comment by MartinThoma** [comment]
See https://users.cs.duke.edu/~tomasi/papers/rubner/rubnerIccv98.pdf

**Comment by WarrenWeckesser** [comment]
[`scipy.stats.wasserstein_distance`](http://scipy.github.io/devdocs/generated/scipy.stats.wasserstein_distance.html) was just [added a few days ago](https://github.com/scipy/scipy/pull/7563).  The [Wasserstein distance](https://en.wikipedia.org/wiki/Wasserstein_metric) is also known as the earth mover's distance.

To any interested devs: should this also be available as one of the metrics in `scipy.spatial.distance`?

**Comment by MartinThoma** [comment]
Nice! I didn't see that.

**Comment by rkern** [comment]
Yeah, I think it's a good candidate for `scipy.spatial.distance`. I'd look for it there when I need it.

**Comment by rgommers** [comment]
From the mailing list (https://mail.python.org/pipermail/scipy-dev/2017-July/022012.html):

*I had a look at the other possible place to put them,
scipy.spatial.distance. While it could fit there as well - your function
signatures fit with distance.cdist - I agree that putting statistical
distances in scipy.stats makes more sense. The Kullback-Leibler divergence
is also present in scipy.stats already (a bit hidden, it's in `entropy`).*

I'd prefer documenting in the ``spatial.distance`` docstring that there are several statistical distances implemented in ``stats`` and cross-link, rather than expose a distance in two different places in the API.

**Comment by MartinThoma** [comment]
Is there any advantage / disadvantage of the two options?

I only see one advantage of document + cross-link: People look at another place and see a group of other distance measures they might be interested in.

(By the way: should I rename the issue to "add Earth mover's distance to `scipy.spatial.distance`?)

**Comment by rgommers** [comment]
> (By the way: should I rename the issue to "add Earth mover's distance to scipy.spatial.distance?)

Sure, always good to have more accurate titles

**Comment by rkern** [comment]
Some external packages point to `scipy.spatial.distance` as a location of distance functions that can be used, and some even let you just give the name and let it do the lookup internally, so not putting it into `scipy.spatial.distance` does have technical costs even if its documentation points to `scipy.stats`.

http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances_argmin.html

**Comment by rgommers** [comment]
@rkern I don't think ``wasserstein_distance`` is suitable for use in things like ``pairwise_distances_argmin``. The former is a metric of how different two probability distributions, and it takes two 1-D arrays and returns a scalar. The latter is point-to-point distances, and wants  ``(n_samples1, n_features)`` and ``(n_samples2, n_features)`` shaped inputs.

That said, it's still possible to move the two functions added in gh-7563 (until we branch 1.0.x) if we decide that makes sense. I'm not in favor of exposing things in two places though.

**Comment by rkern** [comment]
I don't see your point. All of the `scipy.spatial.distance` functions take two 1D vectors and return a scalar. `pairwise_distances_argmin` and company expect that function signature, calling the distance function sequentially with each pair of samples.

**Comment by MartinThoma** [comment]
@rkern Some functions in `scipy.spatial.distance` take more than two parameters (e.g. [seuclidean](http://scipy.github.io/devdocs/generated/scipy.spatial.distance.seuclidean.html#scipy.spatial.distance.seuclidean) and [mahalanobis](http://scipy.github.io/devdocs/generated/scipy.spatial.distance.mahalanobis.html#scipy.spatial.distance.mahalanobis))

**Comment by rkern** [comment]
Sure, but that's not relevant to my point. The typical function signature for the distance functions is exactly what `wasserstein_distance` supports and is what `pairwise_distances_argmin` is expecting.

**Comment by rkern** [comment]
That is, my point was that `pairwise_distances_argmin` isn't expecting a function that takes `(n_samples, n_features)`-shape arrays and spits out an array of distances. It's expecting a function that takes two 1D vectors and spits out a scalar, just like `wasserstein_distance` and the other functions in `scipy.spatial.distance` do. `pairwise_distances_argmin` is doing the pairing itself. `seuclidean` and `mahalanobis` are just a quick `partial()` away from providing the same signature.

**Comment by rgommers** [comment]
@rkern thanks, you have the annoying habit of being right very often:) 

The signatures match. There's two ways to use a scipy distance metric in ``pairwise_distances_argmin``, pass it in as a callable or as a string. The callable way still works for the metrics in ``stats``, the string one won't.

**Comment by tqa236** [comment]
Hi all, thank you very much for the discussion. I would like to know if you're reached a decision over this issue?

**Comment by rkern** [comment]
I'm still in favor of exposing it in `scipy.spatial.distance`.

**Comment by rgommers** [comment]
> I'm still in favor of exposing it in `scipy.spatial.distance`.

That's fine with me. If someone wants to pick this up, that'd be great.

> Hi all, thank you very much for the discussion. I would like to know if you're reached a decision over this issue?

Let's say yes here, thanks for asking.

**Comment by rlucas7** [comment]
Removed the label 'needs-decision', seems from discussion the decision is 'ok' to expose in `scipy.spatial.distance` please comment if incorrect.

**Comment by marcfranquesa** [comment]
Any update on exposing in `scipy.spatial.distance`? @fancidev stated in #21090:

> Undecided whether to symlink into spatial.

Is this the case?

**Comment by rlucas7** [comment]
> Any update on exposing in `scipy.spatial.distance`? [@fancidev](https://github.com/fancidev) stated in [#21090](https://github.com/scipy/scipy/issues/21090):
> 
> > Undecided whether to symlink into spatial.
> 
> Is this the case?

Unless I misunderstand what you are asking, the answer was given by @rgommers with [this comment](https://github.com/scipy/scipy/issues/7804#issuecomment-636341002). 

Much of the discussion on #21090 seems to be about consolidating APIs to make things consistent and some performance improvements. @marcfranquesa is there a comment specifically in the #21090 issue you could link to where it seems to be  unresolved?

**Comment by marcfranquesa** [comment]
I was referring to the second to last sentence in [this comment](https://github.com/scipy/scipy/issues/21090#issuecomment-2202086190), where I read it as it being undecided.

Thanks for the reply and for pointing out the answer in this issue :).
